### PR TITLE
fix(angular): transition animation plays when using browser back and forward buttons

### LINKED ISSUE #16569: [4.0.0-beta.17] Angular's `Location.back()` has no navigation animation
# Bug Report

**Ionic Info**
```
Ionic:

   ionic (Ionic CLI)             : 4.5.0 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.0.0-beta.19
   @angular-devkit/build-angular : 0.10.7
   @angular-devkit/schematics    : 7.0.7
   @angular/cli                  : 7.0.7
   @ionic/angular-toolkit        : 1.2.0

System:

   NodeJS : v10.13.0 (/usr/local/bin/node)
   npm    : 6.4.1
   OS     : macOS Mojave
```

**Describe the Bug**
Using Angular's `Location.back()` has no navigation animation instead of the back navigation animation.

**Steps to Reproduce**
1. Clone this repo on your machine: https://github.com/KevinKelchen/ionic4-stack-page-reuse . It is a slightly modified version of the `sidemenu` Angular stater template. **You will need to switch to the _custom-back-button_ branch**.
2. In the terminal, path into the repo root and run `npm i`.
3. Run `ionic serve`.
4. Open the Chrome DevTools and click the `Toggle device toolbar` button to simulate a mobile device. From the device dropdown choose an iOS device such as an iPhone X.
    - Choosing an iOS device will use the iOS styling mode. This issue is present in both modes; it's just that the iOS styling mode makes the navigation animation direction used more obvious than the default Material Design navigation animation.
4. Click the Menu icon and then `List`.
5. Click `Item 1` to forward navigate to the detail screen for `Item 1`.
6. Click `Item 2` to forward navigate to the detail screen for `Item 2`.
7. Click `Item 3` to forward navigate to the detail screen for `Item 3`.
8. Click `Item 4` to forward navigate to the detail screen for `Item 4`.
9. Notice the in-app Back button.
    - This is _not_ an `ion-back-button` but is rather a similarly styled, custom `app-back-button` component that, when you click it, uses `import { Location } from '@angular/common'` and does a `Location.back()`.
    - The rationale for using a custom Back button instead of `ion-back-button` is stated in #15428 (see [this comment](https://github.com/ionic-team/ionic/issues/15428#issuecomment-438469172)) and is also mentioned in #16516.
10. Click the in-app Back button and watch the navigation animation. You will see as you continue to click the Back button that it uses no navigation animation even though it's using `Location.back()` to go back in the browser history stack.

**Related Code**
A sample application has been provided. See the _Steps to Reproduce_ above.

**Expected Behavior**
I expected that using Angular's `Location.back()` to navigate back in the browser's history stack would use the _back_ navigation animation instead of no navigation animation.

**Additional Context**
See the _Steps to Reproduce_ above for the rationale for using a custom Back button and the related GitHub issues. This behavior change was introduced in Beta 17--perhaps related to fixing #16367.

FWIW, using the Android back button _does_ seem to animate appropriately as does the `ion-back-button`. The browser back/forward buttons have no navigation animation but that seems appropriate given that it's in a browser context.

---

Thanks again for all that you do in making a great framework! 😀

### PR DESCRIPTION
**Pull request by hoi4**
Issue number: resolves #16569

---------

This PR resolves a few issues around navigation transition animations and the navigation stack when using the browser back and forward buttons.

1. The respective transition animation is now applied automatically (without e.g. the workaround described [here](https://github.com/ionic-team/ionic-framework/issues/28176#issuecomment-1720914437))

2. Sometimes when navigating forward with the browser button, pages in the middle of the current tab's stack were incorrectly removed. This caused an issue when toggling between pages via the browser back and forward button (especially visible when using custom transition animations).  

I added some logic to use the current navigation stack as basis for the direction of the animation that will be applied.

Short demonstration video of the changes (using the `ng16` test app):
https://github.com/ionic-team/ionic-framework/assets/23243256/40aba3ec-0c56-44f1-8b82-ba5a2903b414

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

At least to my limited knowledge. The tests are still passing. I just had to increase the timeouts slightly to properly wait for the page transitions to finish.

### CODE REVIEW

**Comment by hoi4** [comment]
@liamdebeasi FYI :)

**Comment by hoi4** [comment]
Any chance to get feedback on these changes? 🙏🏻

**Comment by liamdebeasi** [comment]
Hey there! Sorry for the delay. Someone from the team will do a code review soon. Thanks for contributing!

**Review by liamdebeasi — COMMENTED** [review]
Hey there! Thanks so much for making this PR. I spoke with the team, and we've identified a simpler solution that would address https://github.com/ionic-team/ionic-framework/issues/16569.

The main issue is we explicitly disable animations when using the browser back/forward buttons: https://github.com/ionic-team/ionic-framework/blob/3259da0de181c8f82c38d9de13733213c77d398f/packages/angular/src/providers/nav-controller.ts#L42

`ev.restoredState` is set when the user presses the back/forward buttons: https://angular.io/api/router/NavigationStart#restoredState

Something like the following should resolve the issue:

```diff
if (router) {
      router.events.subscribe((ev) => {
        if (ev instanceof NavigationStart) {
          const id = ev.restoredState ? ev.restoredState.navigationId : ev.id;
-         this.guessDirection = id < this.lastNavId ? 'back' : 'forward';
-         this.guessAnimation = !ev.restoredState ? this.guessDirection : undefined;
+         this.guessAnimation = this.guessDirection = id < this.lastNavId ? 'back' : 'forward';
           this.lastNavId = this.guessDirection === 'forward' ? ev.id : id;
        }
      });
    }
```

------

I know your PR fixes other issues, but we typically don't accept PRs that fix multiple issues at the same time (it makes PRs hard to review). We're happy to review separate PRs for the other issues you worked to resolve in this PR.

Let me know if you have questions!

**Comment by hoi4** [comment]
@liamdebeasi Thank you for your feedback. I understand the issue regarding fixing multiple things simultaneously. I will limit the changes in this PR to the fix for the missing nav animation when using the browser back button. 
I will furthermore create another PR with the rest of my changes.

File: packages/angular/test/base/e2e/src/router-link.spec.ts [hunk 1]
```diff
@@ -113,6 +113,7 @@ describe('Router Link', () => {
   describe('back', () => {
     it('should go back with ion-button[routerLink][routerDirection=back]', () => {
       cy.get('#routerLink-back').click();
+      testBack();
```
Comment (liamdebeasi) [inline packages/angular/test/base/e2e/src/router-link.spec.ts:116]: I don't understand the test change here. The fix modifies whether or not an animation is used, but `testBack` does not verify this.

**Review by liamdebeasi — COMMENTED** [review]
The change itself looks good, but I have a question on the test change.

File: packages/angular/test/base/e2e/src/router-link.spec.ts (hunk 1 above)
Comment (hoi4) [reply packages/angular/test/base/e2e/src/router-link.spec.ts:116]: This was just an issue I fixed where the test executed the back navigation, but did not call `testBack` as the other tests. I assumed it was missed originally when writing the test.

Comment (liamdebeasi) [reply packages/angular/test/base/e2e/src/router-link.spec.ts:116]: Thanks for the clarification. Is this change necessary to validate the bug you fixed? If not, could we remove it from the PR?

Comment (liamdebeasi) [reply packages/angular/test/base/e2e/src/router-link.spec.ts:116]: Thanks for the clarification. Is this change necessary to validate the bug you fixed? If not, could we remove it from the PR?

Comment (liamdebeasi) [reply packages/angular/test/base/e2e/src/router-link.spec.ts:116]: Thanks for the clarification. Is this change necessary to validate the bug you fixed? If not, could we remove it from the PR?

Comment (hoi4) [reply packages/angular/test/base/e2e/src/router-link.spec.ts]: Sure :)

File: packages/angular/test/base/e2e/src/router-link.spec.ts [hunk 2]
```diff
@@ -138,6 +138,7 @@ function testForward() {
   cy.get('app-router-link-page #canGoBack').should('have.text', 'true');
 
   cy.go('back');
+  cy.wait(500);
```
Comment (liamdebeasi) [inline packages/angular/test/base/e2e/src/router-link.spec.ts:141]: Are the timeout changes here necessary? If not, could we revert them? (Sorry meant to comment on this with my last round of reviews, but it slipped my mind)

**Review by liamdebeasi — APPROVED** [review]
The fix itself looks good. Thank you very much! One question below on the test changes.

File: packages/angular/test/base/e2e/src/router-link.spec.ts (hunk 2 above)
Comment (liamdebeasi) [reply packages/angular/test/base/e2e/src/router-link.spec.ts:141]: Also if possible, could you allow me to push changes to your branch? I can resolve the merge conflicts for you.

Comment (hoi4) [reply packages/angular/test/base/e2e/src/router-link.spec.ts:141]: I think, it is necessary. Before, the back transition was instant and now it takes the time for the animation which is why we need the additional `wait`
I invited you to my fork. Thanks for taking care of the merge conflicts.

File: packages/angular/common/src/providers/nav-controller.ts [hunk 3]
```diff
@@ -37,9 +37,9 @@ export class NavController {
     if (router) {
       router.events.subscribe((ev) => {
         if (ev instanceof NavigationStart) {
+          // restoredState is set if the browser back/forward button is used
           const id = ev.restoredState ? ev.restoredState.navigationId : ev.id;
-          this.guessDirection = id < this.lastNavId ? 'back' : 'forward';
-          this.guessAnimation = !ev.restoredState ? this.guessDirection : undefined;
+          this.guessDirection = this.guessAnimation = id < this.lastNavId ? 'back' : 'forward';
```
Comment (liamdebeasi) [inline packages/angular/common/src/providers/nav-controller.ts:42]: One bug I noticed is animations now play when using the back/forward buttons in the `sidemenu` app (clicking items in the menu don't play an animation). I'll see if there's a solution for this.

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
Found one bug, but I'll look into a fix

File: packages/angular/common/src/providers/nav-controller.ts (hunk 3 above)
Comment (liamdebeasi) [reply packages/angular/common/src/providers/nav-controller.ts:42]: Uploading Screen Recording 2023-10-13 at 12.01.33 PM.mov…

Comment (hoi4) [reply packages/angular/common/src/providers/nav-controller.ts:42]: Any news regarding this, @liamdebeasi ?
If it help, I can also take another look at the bug :)

Comment (hoi4) [reply packages/angular/common/src/providers/nav-controller.ts:42]: @liamdebeasi I had a look at the issue. In order to resolve it I needed some of the changes from my other open PR #28297.
I applied the fix to both branches and adapted the test as we don't need the 500ms wait in the `root` case anymore now. Hope that helps!

Comment (hoi4) [reply packages/angular/common/src/providers/nav-controller.ts:42]: @liamdebeasi Any chance you have some time to take a look at the fix? 😇

Comment (liamdebeasi) [reply packages/angular/common/src/providers/nav-controller.ts:42]: I have not. The ticket has been pulled into our current sprint which just started, so I should be able to look at it soon.

**Comment by liamdebeasi** [comment]
Hey there!

I spoke with the team, and we would like to proceed with the fix proposed in https://github.com/ionic-team/ionic-framework/pull/28188#pullrequestreview-1660650092 to resolve the reported issue. I was also able to avoid the need for increased timeouts in the tests through https://github.com/ionic-team/ionic-framework/commit/093c671e3e1733d76ba6760e76af8b934fe4b318.

We looked at the behavior I noted in https://github.com/ionic-team/ionic-framework/pull/28188#discussion_r1358451378 and realized that this was only working by chance. Ionic Angular was never respecting `routerDirection` when using the browser back/forward buttons. Since there was no page transition when using those buttons, the resulting behavior happened to align with the behavior specified in `routerDirection="none"` for the side menu starter app. 

To mitigate any potential impact that this change has, we'd like to release this fix in Ionic 8. I don't have a timeline for Ionic 8, but we are actively working on it. I opened a new PR in https://github.com/ionic-team/ionic-framework/pull/28530 with this fix based off the Ionic 8 development branch. I have also given you co-author credit on the commit for this fix. There is a dev build on the PR so you can try the fix in your application. Note that this dev build is subject to any [Ionic 8 Breaking Changes](https://github.com/ionic-team/ionic-framework/blob/feature-8.0/BREAKING.md#version-8x) (there aren't many at the moment).

I am going to close this in favor of the linked PR, but let me know if you have any questions. Thank you again for contributing this fix!

**Comment by hoi4** [comment]
We have a big release coming up where this change would be very valuable. But I do understand the decision. Looking forward to Ionic 8! 👍🏻

**Pull request closed without merging**